[Inductor][Optimus] Enable pointwise op broadcast to batch with larger size

mengluy0125 opened the pull request:
Summary:
Recently, we observed a lot of mul op is not batched in BAU AFOC 30x model. After looked into its trace, we see a lot of mul ops are not batched due to different shapes, thus we enable the broadcast to improve fusion performance.
https://pxl.cl/4jS6N
Currently, we do not turn on the pass in default since the data copy introduced from repeat may be too expensive to see any fusion benefits.

Test Plan:
# unit test
Buck UI: https://www.internalfb.com/buck2/86308f02-f64a-4247-bd36-e33fecb9428d
Test UI: https://www.internalfb.com/intern/testinfra/testrun/12103424025715455
Network: Up: 13MiB  Down: 144MiB  (reSessionID-7cbeae0b-2e32-4432-a033-724fe2baa70f)
Jobs completed: 263437. Time elapsed: 3:56.1s.
Cache hits: 99%. Commands: 103672 (cached: 103662, remote: 10, local: 0)
Tests finished: Pass 9. Fail 0. Fatal 0. Skip 0. Build failure 0
# local reproduce
```
buck2 run @mode/opt //scripts/jackiexu0313/pt2:local_model_with_pt2 -- --test_mode batch-split --model_type "afoc" --flow_id 529557090
```
P1201029889
```
optimus_scuba_log:  {'before_recompile_pre_grad': 'https://www.internalfb.com/intern/everpaste/?color=0&handle=GB8dhharTWRRo3YDABQwNoEnozImbr0LAAAz', 'BatchLayernormFusion': 'https://www.internalfb.com/intern/everpaste/?color=0&handle=GK9POBQTOvI2Js0LAMlsd1UfgE0kbr0LAAAz', 'normalization_pass_pre_grad': 'https://www.internalfb.com/intern/everpaste/?color=0&handle=GC5r_RY-KRCx5WUDAPMAZ4V7Ojpcbr0LAAAz', 'remove_split_with_size_one_pass_pre_grad': 'https://www.internalfb.com/intern/everpaste/?color=0&handle=GOhNmhawVtn7OVgBAHCJ5lMz4R8mbr0LAAAz', 'after_recompile_pre_grad': 'https://www.internalfb.com/intern/everpaste/?color=0&handle=GAIkmxZxz5TfuRIBADyUBYGAsL4pbr0LAAAz', 'before_recompile_post_grad': 'https://www.internalfb.com/intern/everpaste/?color=0&handle=GKAeuwS_GfWdwyIBAIfmeXP4ko5Qbr0LAAAz', 'PostGradBatchLinearFusion': 'https://www.internalfb.com/intern/everpaste/?color=0&handle=GEAUrhZz0v5mH_MAAN6hfn-7ZHoObr0LAAAz', 'BatchMulPostGradFusion': 'https://www.internalfb.com/intern/everpaste/?color=0&handle=GHFkQxXq6F6Cvl4BAHEPAkIQufdsbr0LAAAz', 'BatchSubPostGradFusion': 'https://www.internalfb.com/intern/everpaste/?color=0&handle=GLRLMBZVkEraBc4CAF4J2rYofBZdbr0LAAAz', 'after_recompile_post_grad': 'https://www.internalfb.com/intern/everpaste/?color=0&handle=GMvpKxfa1CJSzIYBABVGWLcO7A5xbr0LAAAz', 'inductor': Counter({'pattern_matcher_nodes': 1721, 'pattern_matcher_count': 1531, 'normalization_pass': 642, 'remove_split_with_size_one_pass': 629, 'batch_aten_mul_broadcast': 15, 'batch_aten_mul': 13, 'scmerge_split_sections_removed': 11, 'scmerge_cat_removed': 6, 'scmerge_cat_added': 5, 'batch_aten_add_broadcast': 4, 'scmerge_split_removed': 3, 'batch_linear_post_grad': 2, 'batch_aten_sub': 2, 'batch_layernorm': 1}), 'BatchAddBroadcastPostGradFusion': 'https://www.internalfb.com/intern/everpaste/?color=0&handle=GHvyQRkauPU815IBAC13D3y3eOZEbr0LAAAz', 'BatchMulBroadcastPostGradFusion': 'https://www.internalfb.com/intern/everpaste/?color=0&handle=GBTvvhZQAWVuI-IAAMFAdbuxCh5Ibr0LAAAz'}
```

# e2e
training_platform:374b5b38116b4de5e921d3ec3063e19e
### BAU AFOC 30x (3k)
baseline:
f529554280
proposal:
f529596191
### BAU AFOC 30x (2k)
baseline:
f529557090
proposal:
f529597295
### FIRST + CMF
baseline:
f529583511
proposal:
f529583764
### ICVR
baseline:
f529587181
proposal:
f529587775
### IG_CTR
baseline:
f529590319
proposal:
f529590519
### MAI
baseline:
f529591434
proposal:
f529592306
### DPA
baseline:
f529593554
proposal:
f529594305


 {F1454547965}{F1454547968}{F1454547967}
Differential Revision: D53549537




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

yanboliang reviewed (commented):
I'm curious how much gain we can get from this fusion. If the two inputs' tensor size differs a lot, not sure if this brings gain. Probably we should constrain only a few scenarios that we do broadcasting match.

xuzhao9 commented:
Just curious does broadcasting incur additional data copying? Or does it just create a view of the same tensor?

If it incurs additional data copying, it might incur performance overhead that mitigates the benefit of fusion.

mengluy0125 commented on torch/_inductor/fx_passes/group_batch_fusion.py:
```diff
@@ -288,19 +288,46 @@ def __init__(self, op, **kwargs):
+        ), "broadcast -> repeat mapping is not equivalent!"
+
+        return graph.call_function(
+            aten.repeat.default, args=(input_node,), kwargs={"repeats": repeats}
```
Yes. I have run the experiments on MC models, we have 1-3% qps gains observed. Currently, we can have a separate fusion rule with the broadcast, and leave it off in default. Users can try with the pass if they are interested.

jackiexu1992 replied:
Do you have an example when inductor cant fuse?

mengluy0125 replied:
Here is an example ([P1181950842](https://www.internalfb.com/phabricator/paste/view/P1181950842)) from AFOC. As we can see from the patch, many more pointwise tensors could be batched together to fuse if we enable the broadcast.

mengluy0125 commented:
> overhead

Yes, it involves data copy, the overhead depends on the tensor shape to be broadcasted. Based on my experiments, it seems not very bad.

eellison commented on torch/_inductor/fx_passes/group_batch_fusion.py:
```diff
@@ -288,19 +288,46 @@ def __init__(self, op, **kwargs):
-            else False
+        )
+
+    def broadcast_shapes(self, input_shape, other_shape):
+        min_length = min(len(input_shape), len(other_shape))
+        max_length = max(len(input_shape), len(other_shape))
+        if len(input_shape) < len(other_shape):
+            input_shape = [1] * (max_length - min_length) + list(input_shape)
+        else:
+            other_shape = [1] * (max_length - min_length) + list(other_shape)
+        shape_list = []
+        for i in range(max_length):
+            shape_list.append(max(input_shape[i], other_shape[i]))
+
+        return torch.Size(shape_list)
```
you can use `torch.broadcast_shapes` instead

eellison commented on torch/_inductor/fx_passes/group_batch_fusion.py:
```diff
@@ -288,19 +288,46 @@ def __init__(self, op, **kwargs):
+            repeat_out_shape == out_shape
+        ), "broadcast -> repeat mapping is not equivalent!"
+
+        return graph.call_function(
+            aten.repeat.default, args=(input_node,), kwargs={"repeats": repeats}
```
Is there a reason we're using repeat here instead of broadcast_to ? https://github.com/mengluy0125/pytorch/blob/export-D53549537/aten/src/ATen/native/native_functions.yaml#L1359 

or broadcast_tensors ?

mengluy0125 replied:
Thanks for pointing this out. I noticed there are some discussions on how to "expand" the tensor, and chose the repeat one.
https://discuss.pytorch.org/t/torch-repeat-and-torch-expand-which-to-use/27969

eellison commented on torch/_inductor/fx_passes/group_batch_fusion.py:
```diff
@@ -310,7 +337,15 @@ def match(self, node: torch.fx.Node):
             alpha = node.kwargs.get("alpha", 1.0)
```
Would be nice to generalize to get non-Tensor values by looking at the schema instead of hardcoding alpha, and rounding_mode.

mengluy0125 replied:
Can I have more context on schema? Do you mean by looking at its args and kwargs?

eellison replied:
Right, looking at schema args and kwargs. You can call `torch.fx.operator_schemas.normalize_function` with `normalize_to_only_use_kwargs=True` and then hash on all the non-tensor args to generalize the `rounding_mode = node.kwargs.get("rounding_mode", None)` code. Not needed for this pr..

mengluy0125 replied:
In this fusion pass, we combine all pointwise ops, for example, torch.add and torch.mul. The problem is not all pointwise op has the "alpha" and "rounding_mode" keywords. However, in order to generalize to all pointwise ops, we need to assign some "default" values to the ops that do not have them. It seems like normalize_function cannot resolve the hardcode requirements.

eellison reviewed (commented):
New to this code but I would expect a check somewhere of `torch._prims_common.is_expandable_to` somewhere to check that inputs are broadcastable.

mengluy0125 committed:
[Inductor][Optimus] Enable pointwise op broadcast to batch with larger size (#119422)

Summary:

Recently, we observed a lot of mul op is not batched in BAU AFOC 30x model. After looked into its trace, we see a lot of mul ops are not batched due to different shapes, thus we enable the broadcast to improve fusion performance.
https://pxl.cl/4jS6N

Currently, we do not turn on the pass in default since the data copy introduced from repeat may be too expensive to see any fusion benefits.

Test Plan:
# unit test
Buck UI: https://www.internalfb.com/buck2/86308f02-f64a-4247-bd36-e33fecb9428d
Test UI: https://www.internalfb.com/intern/testinfra/testrun/12103424025715455
Network: Up: 13MiB  Down: 144MiB  (reSessionID-7cbeae0b-2e32-4432-a033-724fe2baa70f)
Jobs completed: 263437. Time elapsed: 3:56.1s.
Cache hits: 99%. Commands: 103672 (cached: 103662, remote: 10, local: 0)
Tests finished: Pass 9. Fail 0. Fatal 0. Skip 0. Build failure 0
# local reproduce
```
buck2 run mode/opt //scripts/jackiexu0313/pt2:local_model_with_pt2 -- --test_mode batch-split --model_type "afoc" --flow_id 529557090
```
P1201029889
```
optimus_scuba_log:  {'before_recompile_pre_grad': 'https://www.internalfb.com/intern/everpaste/?color=0&handle=GB8dhharTWRRo3YDABQwNoEnozImbr0LAAAz', 'BatchLayernormFusion': 'https://www.internalfb.com/intern/everpaste/?color=0&handle=GK9POBQTOvI2Js0LAMlsd1UfgE0kbr0LAAAz', 'normalization_pass_pre_grad': 'https://www.internalfb.com/intern/everpaste/?color=0&handle=GC5r_RY-KRCx5WUDAPMAZ4V7Ojpcbr0LAAAz', 'remove_split_with_size_one_pass_pre_grad': 'https://www.internalfb.com/intern/everpaste/?color=0&handle=GOhNmhawVtn7OVgBAHCJ5lMz4R8mbr0LAAAz', 'after_recompile_pre_grad': 'https://www.internalfb.com/intern/everpaste/?color=0&handle=GAIkmxZxz5TfuRIBADyUBYGAsL4pbr0LAAAz', 'before_recompile_post_grad': 'https://www.internalfb.com/intern/everpaste/?color=0&handle=GKAeuwS_GfWdwyIBAIfmeXP4ko5Qbr0LAAAz', 'PostGradBatchLinearFusion': 'https://www.internalfb.com/intern/everpaste/?color=0&handle=GEAUrhZz0v5mH_MAAN6hfn-7ZHoObr0LAAAz', 'BatchMulPostGradFusion': 'https://www.internalfb.com/intern/everpaste/?color=0&handle=GHFkQxXq6F6Cvl4BAHEPAkIQufdsbr0LAAAz', 'BatchSubPostGradFusion': 'https://www.internalfb.com/intern/everpaste/?color=0&handle=GLRLMBZVkEraBc4CAF4J2rYofBZdbr0LAAAz', 'after_recompile_post_grad': 'https://www.internalfb.com/intern/everpaste/?color=0&handle=GMvpKxfa1CJSzIYBABVGWLcO7A5xbr0LAAAz', 'inductor': Counter({'pattern_matcher_nodes': 1721, 'pattern_matcher_count': 1531, 'normalization_pass': 642, 'remove_split_with_size_one_pass': 629, 'batch_aten_mul_broadcast': 15, 'batch_aten_mul': 13, 'scmerge_split_sections_removed': 11, 'scmerge_cat_removed': 6, 'scmerge_cat_added': 5, 'batch_aten_add_broadcast': 4, 'scmerge_split_removed': 3, 'batch_linear_post_grad': 2, 'batch_aten_sub': 2, 'batch_layernorm': 1}), 'BatchAddBroadcastPostGradFusion': 'https://www.internalfb.com/intern/everpaste/?color=0&handle=GHvyQRkauPU815IBAC13D3y3eOZEbr0LAAAz', 'BatchMulBroadcastPostGradFusion': 'https://www.internalfb.com/intern/everpaste/?color=0&handle=GBTvvhZQAWVuI-IAAMFAdbuxCh5Ibr0LAAAz'}
```

# e2e
training_platform:374b5b38116b4de5e921d3ec3063e19e
### BAU AFOC 30x (3k)
baseline:
f529554280
proposal:
f529596191
### BAU AFOC 30x (2k)
baseline:
f529557090
proposal:
f529597295
### FIRST + CMF
baseline:
f529583511
proposal:
f529583764
### ICVR
baseline:
f529587181
proposal:
f529587775
### IG_CTR
baseline:
f529590319
proposal:
f529590519
### MAI
baseline:
f529591434
proposal:
f529592306
### DPA
baseline:
f529593554
proposal:
f529594305


 {F1454547965}{F1454547968}{F1454547967}


# e2e (AFOC)
without broadcast:
f547133320
with broadcast:
f547133027
with both:
f547190976
 {F1476866547}{F1476866543}

Differential Revision: D53549537
